design/src/rfcs/rfc0023_sdk_credential_cache_type_safety.md

```diff
@@ -0,0 +1,260 @@
+}
+
+pub struct EagerCache {
+    credentials_provider: Arc<dyn ProvideCredentials>,
+    // ...
+}
+
+pub struct CustomCache {
+    credentials_provider: Arc<dyn ProvideCredentials>,
+    // Not naming or specifying the custom cache trait for now since its out of scope
+    cache: Arc<dyn SomeCacheTrait>
+}
+
+enum CredentialsCacheInner {
+    Lazy(LazyCache),
+    // Eager doesn't exist today, so this is purely for illustration
+    Eager(EagerCache),
+    // Custom may not be implemented right away
+    Custom(CustomCache),
+}
+
+pub struct CredentialsCache {
+    inner: CredentialsCacheInner,
+}
+
+impl CredentialsCache {
+    // Methods prefixed with `default_` just use the default cache settings
+    pub fn default_lazy(provider: impl ProvideCredentials + 'static) -> Self { /* ... */ }
+    pub fn default_eager(provider: impl ProvideCredentials + 'static) -> Self { /* ... */ }
+
+    // Unprefixed methods return a builder that can take customizations
+    pub fn lazy(provider: impl ProvideCredentials + 'static) -> LazyBuilder { /* ... */ }
+    pub fn eager(provider: impl ProvideCredentials + 'static) -> EagerBuilder { /* ... */ }
+
+    pub(crate) fn create_cache(
+        self,
+        sleep_impl: Arc<dyn AsyncSleep>
+    ) -> SharedCredentialsProvider {
+        // ^ Note: SharedCredentialsProvider would get renamed to SharedCredentialsCache.
+        // This code is using the old name to make it clearer that it already exists,
+        // and the rename is called out in the change checklist.
+        SharedCredentialsProvider::new(
+            match self {
+                Self::Lazy(inner) => LazyCachingCredentialsProvider::new(inner.credentials_provider, settings.time, /* ... */),
+                Self::Eager(_inner) => unimplemented!(),
+                Self::Custom(_custom) => unimplemented!(),
+            }
+        )
+    }
+}
+```
+
+Using a struct over a trait prevents custom caching implementations, but if customization is desired,
+a `Custom` variant could be added to the inner enum that has its own trait that customers implement.
+
+The `SharedCredentialsProvider` needs to be updated to take a cache implementation rather
+than `impl ProvideCredentials + 'static`. A sealed trait could be added to facilitate this.
+
+Configuration would look as follows:
+
+```rust
+let sdk_config = aws_config::from_env()
+    .credentials(CredentialsCache::default_lazy(ImdsCredentialsProvider::builder().build()))
+    .load()
+    .await;
+```
+
+The `credentials_provider` method on `ConfigLoader` would only take `CredentialsCache` as an argument
+so that the SDK could not be configured without credentials caching, or if opting out of caching becomes
+a use case, then a `CredentialsCache::NoCache` variant could be made.
+
+Like Option A, a convenience method can be added to make using the default cache easier:
+
+```rust
+let sdk_config = aws_config::from_env()
+    .credentials_with_default_cache(ImdsCredentialsProvider::builder().build())
```

Comment:
probably wouldn't make the default name longer than custom name—maybe `cache_credentials`? or just `credentials` and document it?